akka23-conductr-client-lib/src/main/scala/com/typesafe/conductr/clientlib/akka/ControlClient.scala

```diff
@@ -0,0 +1,368 @@
+    handler.withConnectedRequest(Payload.bundlesInfo)(handleAsSeq[Bundle])
+
+  /**
+   * @see [[AbstractControlClient.loadBundle()]]
+   * @param bundle The file that is the bundle.
+   *               The filename is important with its hex digest string and is required to be consistent
+   *               with the SHA-256 hash of the bundle’s contents.
+   *               Any inconsistency between the hashes will result in the load being rejected.
+   * @param config Optional: Similar in form to the bundle, only that is the file that describes the configuration.
+   *               Again any inconsistency between the hex digest string in the filename, and the SHA-256 digest
+   *               of the actual contents will result in the load being rejected.
+   * @param cc implicit connection context
+   * @return The result as a Future[BundleRequestResult]. BundleRequestResult is a sealed trait and can be either:
+   *         - BundleRequestSuccess if the loading request has been succeeded. This object contains the request and bundle id
+   *         - BundleRequestFailure if the loading request has been failed. This object contains the HTTP status code and error message.
+   */
+  override def loadBundle(bundle: URI, config: Option[URI])(implicit cc: ConnectionContext): Future[BundleRequestResult] = {
+    def createRequestBody: Future[RequestEntity] = {
+      import cc.actorMaterializer.executionContext
+
+      // TODO: Use Akka stream to extract and read the zip file.
+      //       In the current Akka streams version 2.0.1 there is no utility to extract a zip file and iterate over
+      //       the files inside of the zip file. Therefore a blocking solution inspired by sbt.IO is used.
+      def extractZipEntry(entryName: String, from: URI, destDir: File): Option[URI] = {
+        val zipFile = new File(absolute(from).getPath)
+
+        val zis = new ZipInputStream(new BufferedInputStream(new FileInputStream(zipFile)))
+
+        try {
+          val extractedFile = Stream
+            .continually(zis.getNextEntry)
+            .takeWhile(_ != null)
+            .find(currentEntry => filename(currentEntry.getName) == entryName)
+            .map { entry =>
+              val target = new File(destDir, entryName)
+              val fos = new BufferedOutputStream(new FileOutputStream(target, false))
+
+              try {
+                stream(zis, fos)
+                target
+              } finally {
+                fos.close()
+              }
+            }
+
+          extractedFile.map(_.toURI)
+        } finally {
+          zis.closeEntry()
+          zis.close()
+        }
+      }
+
+      def stream(in: InputStream, out: OutputStream, bufferSize: Int = 8192) = {
+        try {
+          val buffer = new Array[Byte](bufferSize)
+          def read() {
+            val byteCount = in.read(buffer)
+            if (byteCount >= 0) {
+              out.write(buffer, 0, byteCount)
+              read()
```

Comment:
Blocking calls must be surrounded with a `blocking` function e.g. `blocking(read())`, and also use a blocking dispatcher.